### ISSUE TITLE
Modifying in place

### ISSUE THREAD

**Issue by cmaughan** [ToruNiina/toml11#85]
How does mutating existing data work?

I could do:
toml::base loaded = toml::parse(...)
toml::value& table = toml::find(loaded, "mytable");
if (!table.is_initialized())
{
    /* how to set table in loaded *?
   loaded = toml::value {"mytable", {}}
}
The same thing for adding values into existing tables.

The scenario I'm wondering about is loading a file where you don't know the whole contents of the data, but you want to modify a given section and add it if missing....

**Comment by ToruNiina** [comment]
Hi,

When you have a toml file like this,

```toml
[mytable]
key  = "value"
```

you can change the value of `mytable.key` in a way like the following.

```cpp
toml::value file = toml::parse("sample.toml");
toml::value& table = toml::find(file, "mytable");
toml::find(table, "key") = "another value";
```

Here, if a value does not exist, `toml::find` throws an `std::out_of_range` exception. So we need to check a value exists or not before doing this (or catch `std::out_of_range`).

`toml::value` has a member function `as_table()`. By using this, you can gain access to `std::unordered_map<toml::key, toml::value>` that is contained in the `toml::value`.

(Note that when you change the internal container type, like `toml::basic_value<toml::preserve_comments, std::map>`, the return type of `as_table` also changes to the type you set.)

So you can check whether a key exists, append a key-value pair, and do any other operation in the same way as STL after getting a reference to it by calling `as_table()`.

```cpp
toml::value& table = toml::find(file, "mytable");

// check whether "key" exists
if(table.as_table().count("key") != 0)
{
    // access to the value and modify it
    table.as_table().at("key").as_string() += " append";

    // or...
    // toml::find<std::string>(table, "key") += " append";
}
// append or modify a key-value pair in a map anyway
table.as_table()["new-key"] = "new value";
```

**Comment by cmaughan** [comment]
Great, thanks!  Is there a way to make the table & key-value pair and have it overwrite if it is already there?  Your example means I have 2 paths - one where I find and modify, and the other where I create.  I'm wondering if there is a simpler way to force a new value even if it is there; something like:
```
file["mytable"]["new-key"] = "new value";   // Modify or add....
```
?

**Comment by ToruNiina** [comment]
Yes, I know that feature is desired and may be nice to have. Actually, it has already been proposed in #81.

The reason why I didn't implement it in the current release is that I don't have nice idea to make the behaviors of `operator[]`s consistent between arrays and tables.
Some of the STL containers have `operator[]`. In `std::vector`, `std::deque`, `std::array` and raw array, `operator[]` is expected to be ultrafast and unsafe. When an index overruns the range, the behavior is unspecified or undefined. In `std::map` and `std::unordered_map`, `operator[]` always returns a reference to the corresponding value. It makes new value if the corresponding value does not exist.
But `toml::value` can be either. When `toml::value` has a table inside, `operator[]` should make a new value if the corresponding value is not found. When `toml::value` has an array inside, what should happen if an index overruns the range? For the consistency sake, it can make a new value corresponding to the index and return the newly constructed value. But most C++ programmers assumes that every container behaves in a similar way as STL container. `toml::value::operator[]` becomes slower if we check the range and resize an array (but dynamic type checking is unavoidable here). Also, to me, it is not intuitive if `operator[]` for an array appends a new value.

So, in short, I'm concerned about confusing behaviors that appears when I add `operator[]` to `toml::value` in a STL-like way, like

```cpp
toml::value v;
v["key"] = "value"; // OK
v["table"]["foo"] = "bar"; // OK
v["array"][0]     = 42; // Undefined behavior
```

... and complicated and STL-inconsistent behavior of the other option.

```cpp
toml::value v{1,2,3};
v[3] = 42; // OK. appends a value
v[6] = 54; // OK but v[4] and v[5] are uninitialized
           // so serializing this array generates an invalid toml file
```

**Comment by ToruNiina** [comment]
Recently I got independent but identical suggestion about `operator[]`. It convinces me that this feature is strongly desired. So I have implemented it.

Note that I've implemented it in STL-like manner. `v.at(1)` checks if v is an array and its length is larger than 1. Contrary, `v[1]` does not check anything and never throws. If you need type- and bound-checking, use `v.at(n)`.
